src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation.
```

Comment:
I feel a strong need of high-level description of how different parts are supposed to work (with examples using particular numbers).
Especially `QubitListInSharedArray`. There is a number of places, understanding which takes more time than it could, and still no guarantee that the understanding is correct. At least the places where the indices and values are manipulated, like this:
```c++
        sharedQubitStatusArray[lastElement] = id;
        sharedQubitStatusArray[id] = NoneMarker;
```

Reply:
I added comments to the places in QubitListInSharedArray where liked lists are manipulated. I believe it should be easy to understand as this is just a standard singly linked list, managed by a struct named QubitListInSharedArray. Method names are also self-explanatory.
I will add description of the content of the shared array later.